js/private/worker/src/index.ts

```diff
@@ -0,0 +1,82 @@
+import { readWorkRequestSize, writeWorkResponseSize } from './size'
+import { blaze } from './worker_protocol'
+import { Writable } from 'stream'
+
+export type WorkRequest = Omit<
+    ReturnType<typeof blaze.worker.WorkRequest.prototype.toObject>,
+    'cancel'
+> & { signal: AbortSignal; output: Writable }
+
+export type ImplementationFunc = (request: WorkRequest) => Promise<number>
+
+export async function enterWorkerLoop(implementation: ImplementationFunc) {
+    const abortionMap = new Map<number, AbortController>()
+
+    let prev: Buffer = Buffer.alloc(0)
+
+    for await (const buffer of process.stdin) {
+        let chunk: Buffer = Buffer.concat([prev, buffer])
+        let current: Buffer
+
+        const size = readWorkRequestSize(chunk)
+
+        if (size.size <= chunk.length + size.headerSize) {
+            chunk = chunk.slice(size.headerSize)
+            current = chunk.slice(0, size.size)
+            prev = chunk.slice(size.size)
+        } else {
+            prev = chunk
+            continue
+        }
+
+        const request = blaze.worker.WorkRequest.deserialize(current!)
+
+        if (request.cancel) {
+            abortionMap.get(request.request_id)?.abort()
+            continue
+        }
+
+        const abortController = new AbortController()
+        abortionMap.set(request.request_id, abortController)
+
+        const response = new blaze.worker.WorkResponse({
+            request_id: request.request_id,
+        })
+
+        const outputBuffer = Buffer.alloc(0)
+        const outputStream = new Writable({ write: outputBuffer.write })
+
+        implementation({
```

Comment:
no use of await here to achieve a non-blocking loop.  the loop has to be blocked on stdin only.

Reply:
necessary for multiplexing and cancellation

Reply:
👍